lib/IntermediateRepresentation/afforestation.cc

```diff
@@ -0,0 +1,1963 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "afforestation.h"
+#include "builder.h"
+#include "mixin.h"
+#include "../evaluate/fold.h"
+#include "../evaluate/tools.h"
+#include "../parser/parse-tree-visitor.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace Fortran::IntermediateRepresentation {
+
+static llvm::raw_ostream *debugChannel;
+static llvm::raw_ostream &DebugChannel() {
+  return debugChannel ? *debugChannel : llvm::errs();
+}
+static void SetDebugChannel(llvm::raw_ostream *output) {
+  debugChannel = output;
+}
+void SetDebugChannel(const std::string &filename) {
+  std::error_code ec;
+  SetDebugChannel(
+      new llvm::raw_fd_ostream(filename, ec, llvm::sys::fs::F_None));
+  CHECK(!ec);
+}
+
+namespace {
+struct LinearOp;
+
+using LinearLabelRef = unsigned;
+constexpr LinearLabelRef unspecifiedLabel{~0u};
+
+struct LinearLabelBuilder {
+  LinearLabelBuilder() : referenced(32), counter{0u} {}
+  LinearLabelRef getNext() {
+    LinearLabelRef next{counter++};
+    auto cap{referenced.capacity()};
+    if (cap < counter) referenced.reserve(2 * cap);
+    referenced[next] = false;
+    return next;
+  }
+  void setReferenced(LinearLabelRef label) { referenced[label] = true; }
+  bool isReferenced(LinearLabelRef label) const { return referenced[label]; }
+  std::vector<bool> referenced;
```

Comment:
`std::set<int>` might be safer with respect to expansion and more readable (since this is actually meant to represent a set).  The extra overhead won't matter.